### DISCUSSION TITLE
Is there a way to build a single zip (tar) file with multiple products (configurations) in it?

### DISCUSSION THREAD

**Discussion by jonahgraham** [eclipse-tycho/tycho#3076] [Q&A]
**Question:** Is there a way to build a single zip of an Eclipse install that can be launched in two different ways, i.e. have two different configurations (driven by two different product files)?

**Background**

The Eclipse CDT project has a feature called the [Standalone Debugger](https://github.com/eclipse-cdt/cdt/blob/main/StandaloneDebugger.md) that allows users to launch their Eclipse install in a very UI limited way. This is achieved by having a [product](https://github.com/eclipse-cdt/cdt/blob/4aa793a6515f30fbb92090e2431dc7132c5cb971/debug/org.eclipse.cdt.debug.application.product/debug.product) file that we can build that contains the limited set of bundles (features) that we want. It is basically a RCP product (e.g. no perspectives, as few UI elements as possible that are not debug flow related).

Eclipse CDT is also delivered through the EPP CPP package, which is the "full fat" version for developing and debugging.

For a long time we have [delivered a config.ini](https://github.com/eclipse-cdt/cdt/blob/3bcba2303d512b31b6587e6eba59a0221a824f56/debug/org.eclipse.cdt.debug.application/pom.xml#L34-L45) that lists just the bundles we want when users want to launch the Standalone Debugger (we [start eclipse](https://github.com/eclipse-cdt/cdt/blob/4aa793a6515f30fbb92090e2431dc7132c5cb971/debug/org.eclipse.cdt.debug.application/scripts/cdtdebug.sh#L104-L107) with `-config` pointing at the `config.ini` in a shell script). 

Managing this has become increasingly difficult and is prone to errors as the contents of the config.ini are not expressed in any of the p2 metadata and therefore when EPP CPP is built, there is no way to guarantee that all the bundles that existed when CDT was built will be installed. This has become substantially more error prone with the modernization of third-party code (mostly due to churn).

Some solution is needed to resolve https://github.com/eclipse-cdt/cdt/issues/591 to keep having the standalone debugger.

**Comment by mickaelistria** [comment]
I don't think it's a Tycho-specific issue, but more a p2 one.
IIRC, p2 can allow multiple *profiles* for a given install location; which allow to share some things such as the bundle pool, so the bundle exist only once.
However, if the "Standalone Debugger" is a subset of the whole CDT, I don't think you need a product, but just the application; so the same product installation can server the IDE and the standalone debugger as distinct applications. You would then only need to provide some shortcut scripts to launch the standalone debugger instead of the IDE (passing the proper `-application` flag to the launcher). Or are there some product configuration required in the standalone debugger you can't put in the CDT product?

**Comment by laeubi** [comment]
Isn't that something a custom perspective can be used for?
If you just need the configuration files, you can also build two distinct products and then copy over from one to the other using the [copy-resources mojo](https://maven.apache.org/plugins/maven-resources-plugin/examples/copy-resources.html).

But there is one thing I don't understand you wrote 

> there is no way to guarantee that all the bundles that existed when CDT was built will be installed

So does the standalone needs *additional* bundles?

Beside that, I think Oomph might be able to build a derived product from a running eclipse (@merks ?) so you could add a button to CDT that performs this inside Eclipse and just ask the user where to place the standalone debugger (assuming you don't want to consider to distribute two products in the first place).
